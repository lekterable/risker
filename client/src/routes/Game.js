import React, { Component } from 'react'
import { notify, endGame } from '../actions'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

class Game extends Component {
	constructor(props) {
		super(props)
		if (!this.props.playing) {
			this.props.notify('error', 'You are not in any game!')
			return this.props.history.push('/')
		}
	}

	state = {
		turn: '',
		host: '',
		player: {
			total: 0,
			round: 0,
			roll: [0, 0]
		},
		opponent: {
			total: 0,
			round: 0,
			roll: [0, 0]
		}
	}

	roll = () => this.props.socket.emit('turn-roll')

	end = () => this.props.socket.emit('turn-end')

	componentDidMount() {
		this.props.socket.emit('ready', game => {
			if (!game && this.props.playing) {
				this.props.endGame()
				this.props.notify('info', 'Game has ended!')
			}
			if (!game) {
				return this.props.history.push('/')
			}
			const player = game.players.find(
				player => player.id === this.props.socket.id
			)
			const opponent = game.players.find(
				player => player.id !== this.props.socket.id
			)
			this.setState({
				turn: game.turn,
				host: game.host,
				player: { ...player },
				opponent: { ...opponent }
			})
		})
		this.props.socket.on('update', game => {
			const player = game.players.find(
				player => player.id === this.props.socket.id
			)
			const opponent = game.players.find(
				player => player.id !== this.props.socket.id
			)
			this.setState({
				turn: game.turn,
				host: game.host,
				player: { ...player },
				opponent: { ...opponent }
			})
		})
		this.props.socket.on('win', game => {
			const player = game.players.find(
				player => player.id === this.props.socket.id
			)
			const opponent = game.players.find(
				player => player.id !== this.props.socket.id
			)
			this.setState({
				turn: game.turn,
				host: game.host,
				player: { ...player },
				opponent: { ...opponent }
			})
			console.log(
				game.players.find(player => player.total + player.round >= 100)
			)
		})
	}

	componentWillUnmount() {
		this.props.socket.off('update')
		this.props.socket.off('win')
	}

	render() {
		return (
			<div>
				<div className="row">
					<div className="col-12 col-md-6 text-center">
						<div>Your total score: {this.state.player.total}</div>
						<div>Your round score: {this.state.player.round}</div>
						<div>
							Your roll:
							<span role="img" aria-label="dice">
								ðŸŽ²
							</span>
							{this.state.player.roll[0]} and
							<span role="img" aria-label="dice">
								ðŸŽ²
							</span>
							{this.state.player.roll[1]}
						</div>
					</div>
					<div className="col-12 col-md-6 text-center">
						<div>Opponent's total score: {this.state.opponent.total}</div>
						<div>Opponent's round score: {this.state.opponent.round}</div>
						<div>
							Opponent's roll:
							<span role="img" aria-label="dice">
								ðŸŽ²
							</span>
							{this.state.opponent.roll[0]} and
							<span role="img" aria-label="dice">
								ðŸŽ²
							</span>
							{this.state.opponent.roll[1]}
						</div>
					</div>
				</div>
				<div className="row">
					<div className="col-12 col-md-6 text-center">
						<button
							className="btn btn-outline-primary"
							disabled={this.props.socket.id !== this.state.turn}
							onClick={this.roll}
						>
							Roll
						</button>
						<button
							className="btn btn-outline-primary"
							disabled={this.props.socket.id !== this.state.turn}
							onClick={this.end}
						>
							End
						</button>
					</div>
				</div>
			</div>
		)
	}
}

export default connect(
	state => ({ ...state }),
	dispatch => ({
		notify: (type, message, title) => dispatch(notify(type, message, title)),
		endGame: () => dispatch(endGame())
	})
)(withRouter(Game))
