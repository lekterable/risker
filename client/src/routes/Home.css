.invite-form {
	margin-top: 40px;
}
.players-list {
	margin-top: 50px;
}
