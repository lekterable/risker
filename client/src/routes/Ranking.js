import React, { Component } from 'react'
import { connect } from 'react-redux'

class Ranking extends Component {
  render(){
    const players = ['player1', 'player2', 'player3', 'player4'].map((player, index) => <li key={index}>{player}</li>)
    return (
      <div>
        {players}
      </div>
    )
  }
}
const mapStateToProps = (state) => ({...state})
export default connect(mapStateToProps)(Ranking)