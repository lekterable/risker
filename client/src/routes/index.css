.footer {
  margin-top: 40px;
}
nav {
  margin-bottom: 40px;
}