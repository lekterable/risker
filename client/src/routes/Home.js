import React, { Component } from 'react'
import { connect } from 'react-redux'
import { startGame, endGame, notify } from '../actions'
import { withRouter } from 'react-router-dom'
import './Home.css'

class Home extends Component {
  state = {
    players: [],
    opponent: ''
  }
  invite = () => {
    this.props.startGame(this.props.socket.id)
    this.props.socket.emit('invitation', this.state.opponent, (res) => res?this.handleAcceptedInvite():this.handleDeclinedInvite())
  }
  handleAcceptedInvite = () => {
    this.props.notify('success', 'Accepted')
  }
  handleDeclinedInvite = () => {
    this.props.notify('error', 'Declined')
  }
  acceptInvite = (room, callback) => {
    callback(true)
    this.props.startGame(room)
  }
  declineInvite = (room, callback) => {
    callback(false)
    this.props.endGame(room)
  }
  componentDidMount() {
    this.props.socket.emit('info')

    this.props.socket.on('start-game', ()=>{
      this.props.history.push('/game')
    })

    this.props.socket.on('info', (data) => {
      this.setState({players: data.players.filter((player) => player !== this.props.socket.id)})
    })
    
    this.props.socket.on('invitation', (data, callback) => {
      if(this.props.playing)
        return callback(false)
      this.props.notify('info', <span><button className='btn btn-primary btn-sm' onClick={() => this.acceptInvite(data, callback)}>Accept</button><button className='btn btn-primary btn-sm' onClick={() => this.declineInvite(data, callback)}>Decline</button></span>,'Received an invitation!')
    })
    
  }
  componentWillUnmount() {
    this.props.socket.off('start-game')
    this.props.socket.off('info')
    this.props.socket.off('invitation')
  }
  render(){
    const players = this.state.players.map((player, index) => <li className='list-group-item text-center' key={index}>{player}</li>)
    return (
      <div className='main'>
        <div className='form-group row'>
          <label className='col-3 offset-md-3 col-md-2 offset-lg-4 col-lg-1 col-form-label text-center'>My ID:</label>
          <div className='col-9 col-md-5 col-lg-3'><input className='form-control text-center' type='text' readOnly value={this.props.socket.id || ''}/></div>
        </div>
        <div className='players-list row'>
          <div className='text-center col-12'>Players:</div>
          <div className='col-12 col-md-6 offset-md-3 col-lg-4 offset-lg-4'>
            <ul className='list-group'>
              {players}
            </ul>
          </div>
        </div>
        <div className='invite-form row'>
          <div className='col-12 text-center'>
            <input className='form-control text-center col-8 offset-2 col-md-4 offset-md-4' placeholder='Insert ID' type='text' value={this.state.opponent} onChange={(e) => this.setState({opponent: e.target.value})}/>
            <button className='btn btn-outline-primary' onClick={() => this.props.socket.id !== this.state.opponent?this.invite():''}>Invite</button>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({...state})
const mapDispatchToProps = (dispatch) => ({startGame: (room) => dispatch(startGame(room)), notify: (type, message, title) => dispatch(notify(type, message, title)), endGame: (room) => dispatch(endGame(room))})
export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Home))