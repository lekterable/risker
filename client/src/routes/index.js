import React, { Component } from 'react'
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import { connect } from 'react-redux'
import { connectSocket, createNotifier } from '../actions'
import { NotificationContainer, NotificationManager } from 'react-notifications'
import 'react-notifications/lib/notifications.css'
import io from 'socket.io-client'
import Header from '../components/Header'
import Footer from '../components/Footer'
import Home from './Home'
import Ranking from './Ranking'
import Game from './Game'
import './index.css'

class App extends Component {
  constructor(props) {
    super(props)
    this.props.connectSocket(io.connect('http://localhost:3001'))
    this.props.createNotifier(NotificationManager)
  }
  render() {
    return (
      <div>
        <Router>
          <div className="container App">
            <Header />
            <NotificationContainer />
            <Switch>
              <Route exact path="/ranking" render={() => <Ranking />} />
              <Route exact path="/game" render={() => <Game />} />
              <Route render={() => <Home />} />
            </Switch>
            <Footer content="Made with â™¥ by github.com/lekterable" />
          </div>
        </Router>
      </div>
    )
  }
}

const mapStateToProps = state => ({ socket: state.socket })
const mapDispatchToProps = dispatch => ({
  connectSocket: socket => dispatch(connectSocket(socket)),
  createNotifier: manager => dispatch(createNotifier(manager))
})
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
