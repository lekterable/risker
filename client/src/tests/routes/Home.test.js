import React from 'react'
import { shallow } from 'enzyme'
import { Home } from '../../routes/Home'

describe('Home', () => {
  const props = (props, socketProps) => ({
    socket: {
      id: '',
      emit: () => {},
      on: () => {},
      ...socketProps
    },
    ...props
  })

  it('should match snapshot', () => {
    const wrapper = shallow(<Home {...props()} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should display player id correctly', () => {
    const wrapper = shallow(<Home {...props({}, { id: 'player-id' })} />)

    expect(wrapper.find('input.player-id').prop('value')).toEqual('player-id')
  })

  it('should call `emit` on mount', () => {
    const emitSpy = jest.fn()

    const wrapper = shallow(<Home {...props({}, { emit: emitSpy })} />)

    expect(emitSpy.mock.calls.length).toEqual(1)
  })

  it('should call `on` on mount', () => {
    const onSpy = jest.fn()

    const wrapper = shallow(<Home {...props({}, { on: onSpy })} />)

    expect(onSpy.mock.calls.length).toEqual(3)
    expect(onSpy.mock.calls[0][0]).toEqual('start-game')
    expect(onSpy.mock.calls[1][0]).toEqual('info')
    expect(onSpy.mock.calls[2][0]).toEqual('invitation')
  })

  it('should set opponent id on input', () => {
    const wrapper = shallow(<Home {...props()} />)
    wrapper.find('input.opponent-id').simulate('change', {
      target: {
        value: 'opponent'
      }
    })

    expect(wrapper.state('opponent')).toEqual('opponent')
  })

  it('should not call `invite` on button click if no opponent id provided', () => {
    const inviteSpy = jest.fn()

    const wrapper = shallow(<Home {...props({}, { id: 'player-id' })} />)
    wrapper.instance().invite = inviteSpy
    wrapper.find('button').simulate('click')

    expect(inviteSpy.mock.calls.length).toEqual(0)
  })

  it('should not call `invite` on button click if id belongs to the player', () => {
    const inviteSpy = jest.fn()

    const wrapper = shallow(<Home {...props({}, { id: 'player-id' })} />)
    wrapper.instance().invite = inviteSpy
    wrapper.setState({ opponent: 'player-id' })
    wrapper.find('button').simulate('click')

    expect(inviteSpy.mock.calls.length).toEqual(0)
  })

  it('should call `invite` on button click', () => {
    const inviteSpy = jest.fn()

    const wrapper = shallow(<Home {...props({}, { id: 'player-id' })} />)
    wrapper.instance().invite = inviteSpy
    wrapper.setState({ opponent: 'opponent-id' })
    wrapper.find('button').simulate('click')

    expect(inviteSpy.mock.calls.length).toEqual(1)
  })

  it('should render player list correctly', () => {
    const players = ['opponent-1', 'opponent-2', 'opponent-3']

    const wrapper = shallow(<Home {...props()} />)

    expect(wrapper.find('li.list-group-item').length).toEqual(0)

    wrapper.setState({ players })

    expect(wrapper.find('li.list-group-item').length).toEqual(3)
  })

  it('should set opponent id on click', () => {
    const players = ['opponent-1', 'opponent-2', 'opponent-3']

    const wrapper = shallow(<Home {...props()} />)
    wrapper.setState({ players })
    wrapper
      .find('li.list-group-item')
      .at(0)
      .simulate('click')

    expect(wrapper.state('opponent')).toEqual('opponent-1')
  })
})
