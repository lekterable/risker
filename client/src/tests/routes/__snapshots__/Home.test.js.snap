// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
<div
  className="main"
>
  <div
    className="form-group row"
  >
    <label
      className="col-3 offset-md-3 col-md-2 offset-lg-4 col-lg-1 col-form-label text-center"
    >
      My ID:
    </label>
    <div
      className="col-9 col-md-5 col-lg-3"
    >
      <input
        className="form-control text-center player-id"
        readOnly={true}
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="players-list row"
  >
    <div
      className="text-center col-12"
    >
      Players:
    </div>
    <div
      className="col-12 col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <ul
        className="list-group"
      />
    </div>
  </div>
  <div
    className="invite-form row"
  >
    <div
      className="col-12 text-center"
    >
      <input
        className="form-control text-center col-8 offset-2 col-md-4 offset-md-4 opponent-id"
        onChange={[Function]}
        placeholder="Insert ID"
        type="text"
        value=""
      />
      <button
        className="btn btn-outline-primary"
        onClick={[Function]}
      >
        Invite
      </button>
    </div>
  </div>
</div>
`;
